### PR TITLE
test_models: add tx fuzzy test

### PR DESCRIPTION
**Pull request by bongbui321**
resolve [#32425](https://github.com/commaai/openpilot/issues/32425)

Uses `FuzzyGenerator` to generate fuzzy `CarControl` messages

hypothesis example generation is extremly slow. 

Nissan speed: 4 wheels in OP and 2 rear wheels in panda
Nissan steering angle out of -1311 - 1310 -> CANPacker wrong value
Tesla has clip to current steering angle

### CODE REVIEW

**Comment by sshane** [comment]
Let me know when you want a review!

**Comment by adeebshihadeh** [comment]
Nice work - it's a great sign that it's finding bugs!

**Comment by adeebshihadeh** [comment]
Bounty locked!

> hypothesis example generation is extremly slow.

Might be worth looking into https://github.com/commaai/openpilot/issues/32693

**Comment by bongbui321** [comment]
yeah, ideally, I want to create a list of CarControl messages rather than only one such as now, it takes nearly 18min on my machine to finish half the tx_fuzzy test. 

> Might be worth looking into https://github.com/commaai/openpilot/issues/32693

Our current Hypothesis is already the faster version, the latest release of hypothesis is 2x slower :).

**Comment by bongbui321** [comment]
Might merge tx fuzzy with carstate fuzzy.

Occasionally catches https://github.com/commaai/panda/pull/1948. Not entirely sure why hypothesis doesn't consistenly repro that failed data generation

**Comment by bongbui321** [comment]
@sshane ready for review

File: selfdrive/car/tests/test_models.py [hunk 1]
```diff
@@ -323,10 +329,13 @@ def test_car_controller(car_control):
   @settings(max_examples=MAX_EXAMPLES, deadline=None,
             phases=(Phase.reuse, Phase.generate, Phase.shrink))
   @given(data=st.data())
-  def test_panda_safety_carstate_fuzzy(self, data):
+  def test_panda_safety_carstate_tx_fuzzy(self, data):
```
[the remark this answers is not in the document]
Comment (adeebshihadeh) [reply selfdrive/car/tests/test_models.py:332]: oh is this merged for speed?

**Review by adeebshihadeh — COMMENTED** [review]
I don't think we merged all your bugfixes yet, so how is this passing? Are we not getting good coverage on a single run?

If so, what would it take to hit all the known bugs on a single run? How long would that take in CI?

**Comment by bongbui321** [comment]
> I don't think we merged all your bugfixes yet, so how is this passing? Are we not getting good coverage on a single run?

I don't think we merged all your bugfixes yet, so how is this passing? Are we not getting good coverage on a single run?

Nissan max steering angle `[issue](https://github.com/commaai/openpilot/issues/32978)` is opened and included in this PR

Tesla `[issue](https://github.com/commaai/openpilot/issues/32999)`: This might happen with a few runs, and I suspect that if run locally, `hypothesis` will try to recreate that failing examples pretty consistently. That requires a more thoughtful fix, which I'm hesitant to open a fix PR since I'm not familiar with the expected behaviour of the platform

The Nissan speed issue is addressed here: 
```python3
    # Nissan calculate vEgoRaw with avg speed of 4 wheels
    # while panda get vehicle_speed from rear wheels only,
    # so there might be a big enough difference between the two
    if self.CP.carName == "nissan":
      panda_vehicle_speed_last = self.safety.get_vehicle_speed_last() / VEHICLE_SPEED_FACTOR
      if abs(self.CI.CS.out.vEgoRaw - panda_vehicle_speed_last) > 0.2:
        cs_msg = self.CI.CS.out.to_dict()
        cs_msg["vEgoRaw"] = panda_vehicle_speed_last
        self.CI.CS.out = car.CarState.new_message(**cs_msg).as_reader()
```
This realistically shouldn't happen front wheels and rear wheel speed should be close

**Comment by adeebshihadeh** [comment]
Can you rebase?

**Comment by maxime-desroches** [comment]
This PR was closed because of a git history rewrite.
Please read https://github.com/commaai/openpilot/issues/33399 for what to do with your fork and your PRs.

**Pull request closed without merging**